pypesto/collections/collections.py

```diff
@@ -0,0 +1,497 @@
+import numpy as np
+import pandas as pd
+from typing import Sequence, Tuple, Callable, Dict
+
+from ..prediction import PredictionResult, PredictionConditionResult
+from .constants import (PREDICTOR, PREDICTION_ID, PREDICTION_RESULTS,
+                        PREDICTION_ARRAYS, PREDICTION_SUMMARY, OUTPUT,
+                        OUTPUT_SENSI, TIMEPOINTS, X_VECTOR, NX, X_NAMES,
+                        NVECTORS, VECTOR_TAGS, PREDICTIONS, MODE_FUN,
+                        CollectionType, COLLECTION_TYPE, MEAN, MEDIAN,
+                        STANDARD_DEVIATION, PERCENTILE, SUMMARY, LOWER_BOUND,
+                        UPPER_BOUND)
+
+
+class CollectionPrediction:
+    """
+    A collection prediction consists of a collection, i.e., a set of parameter
+    vectors and their identifiers such as a sample or an ensemble, and a
+    prediction function. It can be attached to a collection-type object
+    """
+
+    def __init__(self,
+                 predictor: Callable,
+                 prediction_id: str = None,
+                 prediction_results: Sequence[PredictionResult] = None,
+                 lower_bound: Sequence[np.ndarray] = None,
+                 upper_bound: Sequence[np.ndarray] = None):
+        """
+        Constructor.
+
+        Parameters
+        ----------
+        predictor:
+            Prediction function, e.g., an AmiciPredictor
+        prediction_id:
+            Identifier for the predictions
+        prediction_results:
+            List of Prediction results
+        lower_bound:
+            Array of potential lower bounds for the predictions, should have
+            the same shape as the output of the predictions, i.e., a list of
+            numpy array (one list entry per condition), with the arrays having
+            the shape of n_timepoints x n_outputs for each condition.
+        upper_bound:
+            array of potential upper bounds for the parameters
+        """
+        self.predictor = predictor
+        self.prediction_id = prediction_id
+        self.prediction_results = prediction_results
+        if prediction_results is None:
+            self.prediction_results = []
+
+        # handle bounds
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+
+        self.prediction_arrays = None
+        self.prediction_summary = {MEAN: None,
+                                   STANDARD_DEVIATION: None,
+                                   MEDIAN: None}
+
+    def __iter__(self):
+        yield PREDICTOR, self.predictor
+        yield PREDICTION_ID, self.prediction_id
+        yield PREDICTION_RESULTS, self.prediction_results
+        yield PREDICTION_ARRAYS, self.prediction_arrays
+        yield PREDICTION_SUMMARY, {i_key: dict(self.prediction_summary[i_key])
+                                   for i_key in self.prediction_summary.keys()}
+        yield LOWER_BOUND, self.lower_bound
+        yield UPPER_BOUND, self.upper_bound
+
+    def condense_to_arrays(self):
```

Comment:
```suggestion
    def _condense_to_arrays(self):
```
Not sure, but maybe something to be considered as private? (Also check for others)

Reply:
no necessarily private, also tested (to be called) from the outside.